python-packaging/execute-code.md

```diff
@@ -0,0 +1,167 @@
+  - don't have to remember which
+- don't have to use the `python` command
+  - don't have to even remember it is a Python script
+
+
+# Share Code
+
+## Execute a python package
+
+In [Code Workflow Logic][Code Workflow Logic] you learned of the two primary ways to execute a stand-alone Python script.
+There are two other ways to execute Python as commands, both of which work for code that has been formatted as a package.
+
+### Entrypoints
+
+There is a special `entrypoint` a package can specify in its configuration which will direct installers to create an
+executable command. Entrypoints are a general purpose plug-in system for Python packages, but the
+[`console_scripts`](https://packaging.python.org/en/latest/specifications/entry-points/#use-for-scripts)
+entry is specifically targeted at creating executable commands on systems that install the package.
+
+The target of a `scripts` definition should be one function within your package, which will be directly executed
+when the command is invoked in the shell. A `scripts` definition in your `pyproject.toml` looks like:
+
+```toml
+[project.scripts]
+COMMAND = "my_package.my_module:my_function"
+```
+
+where `COMMAND` is the name of the command that will be made available after installation, `my_package` is the name of
+your top-level package import, `my_module` is the name of any sub-modules in your package (optional, or may be
+repeated as necessary to access the correct sub-module), and `my_funciton` is the function that will be called
+(without parameters) when the command is invoked.
+
+Scripts defined in project configuration, such as `pyproject.toml`, do not need to exist as independent files in
+the package repository, but will be created by installation tools, such as `pip`, at the time the package is
+installed, in a manner customized to the current operating system.
+
+### Executable modules
+
+The final way to make Python code executable directly from the command line is to include a
+[`__main__` module](https://docs.python.org/3/library/__main__.html#module-__main__) in your package. Any package that
+contains a `__main__` module and is installed in the current Python environment can be execute as a module
+directly from the `python` command, without reference to any specific files.
+```
+python -m my_package
+```
+
+Try to create a `__main__.py` module in your package that will execute with the above command. (don't forget to
+(re)install your package after creating this file!)
+
+#### Further exploration
+
+On your own or in small groups:
+
+- What might be the advantages of making a packaged executable over providing script entry points?
+- What are some disadvantages?
+- Review the Pros section from [Executing Scripts][Executing Scrips]
+  - Any similarities between executable packages and executable scripts?
+
+#### More About Main
+
+You just learned that the `__main__` module allows a package to be executed directly from the command line with
```

Comment:
i was wondering where this was, ha :)

I think the flow might be a little clearer if we go
- execute single python files with `python file.py`
- execute a file within a python package with `python -m mypackage.myfile` initially with some loose calls at the module level, and then demonstrating why we might want to wrap those up in a function or put them in a `if __name__ == "__main__":` block so they don't run when imported,
- execute a special `__main__.py` file that takes the place of that `if __name__ == "__main__":` block

Just because I think introducing a new dunder file might take some motivating, like "why do i need this special file" and so introducing directly running a module file first gets you that motivation and also is a nice bridge from running a python script directly.

Reply:
Hmm, this is actually part of what I am trying to teach here. `python -m` is _not_ for single-file execution, it only works to execute (part of) a package. So you can't introduce `-m` without `__main__.py` plus a full intro to packaging (building, pyproject.toml, installing, ...). Also `__main__.py` files typically don't have a `if __name__ == "__main__":` because their purpose is only to be executed, not imported. (this is called out in the standard documentation).

So this may need some reordering, but you seem to have dropped entrypoints in your suggetion, and merged `python file.py and `./file.py` (very understandable). I do think that the `if` guard could maybe come first, as it is IME more common than `__main__.py`

Reply:
I like this a lot. I took most of this approach in my next draft.